image/image.go

```diff
@@ -172,6 +179,32 @@ func (i *genericImage) getSchema1Manifest() (manifest, error) {
 	return mschema1, nil
 }
 
+// uniqueLayerDigests returns a list of layer digests referenced from a manifest.
+// The list will not contain duplicates; it is not intended to correspond to the "history" or "parent chain" of a Docker image.
+func uniqueLayerDigests(m manifest) []string {
+	var res []string
+	seen := make(map[string]struct{})
+	for _, digest := range m.GetLayers() {
+		if _, ok := seen[digest]; ok {
+			continue
+		}
+		seen[digest] = struct{}{}
+		res = append(res, digest)
+	}
+	return res
+}
+
+// LayerDigests returns a list of layer digests referenced by this image.
+// The list will not contain duplicates; it is not intended to correspond to the "history" or "parent chain" of a Docker image.
+// NOTE: It is essential for signature verification that LayerDigests is computed from the same manifest which is returned by Manifest().
+func (i *genericImage) LayerDigests() ([]string, error) {
+	m, err := i.getSchema1Manifest()
```

Comment:
I think that would happen somewhere around `genericImage.Manifest` (note that `getSchema1Manifest already calls `.Manifest`)—before we ever return anything manifest-dependent to any user of this object, we need to decide which MIME version to use; and then we commit to that decision.

You have already built the `manifest` interface as an abstraction, so I guess `getSchema1Manifest` would end up a `getParsedManifest` which returns an object implementing `manifest` for whatever MIME version was selected as described above.  Then both `LayerDigests` and `Inspect` would use this `manifest` implementation.

All this requires deciding what to do for the non-v2s1 data returned by `Inspect` and `LayersCommand`; we will need to do that eventually, there’s no avoiding that.